### PR TITLE
[GSK-3365] Ease custom metrics support

### PR DESCRIPTION
**Pull request by pierlj**

### CODE REVIEW

File: giskard/datasets/metadata/text_metadata_provider.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def _detect_charset(text: str):
     return charset or "undefined"
 
 
-def _detect_lang(text: str):
+def detect_lang(text: str):
```
Comment (mattbit) [inline giskard/datasets/metadata/text_metadata_provider.py:38]: maybe worth moving this out, in a shared module

File: giskard/rag/evaluate.py [hunk 2]
```diff
@@ -76,17 +80,27 @@ def evaluate(
     # @TODO: improve this
     metrics = metrics or []
     if not any(isinstance(metric, CorrectnessMetric) for metric in metrics):
-        metrics.append(CorrectnessMetric(name="Correctness", agent_description=agent_description))
+        # By default only correctness is computed as it is required to build the report
+        metrics.insert(
```
Comment (mattbit) [inline giskard/rag/evaluate.py:84]: just noticed this: `metrics` needs to be a `list` or casted with `list(metrics)`, not a generic sequence

File: giskard/rag/evaluate.py [hunk 3]
```diff
@@ -76,17 +80,27 @@ def evaluate(
...
+            0, CorrectnessMetric(name="Correctness", llm_client=llm_client, agent_description=agent_description)
+        )
+
+    metrics_results = defaultdict(dict)
 
-    metrics_results = {}
     for metric in metrics:
-        metrics_results.update(metric(testset, answers, llm_client=llm_client))
+        metric_name = metric.name if isinstance(metric, Metric) else metric.__name__
```
Comment (mattbit) [inline giskard/rag/evaluate.py:91]:
```
getattr(metric, "name", metric.__name__)
```

File: giskard/rag/evaluate.py [hunk 4]
```diff
@@ -15,6 +16,9 @@
 
 logger = logging.getLogger(__name__)
 
+REPORT_CORRECTNESS_COL = "correctness"
```
Comment (mattbit) [inline giskard/rag/evaluate.py:19]: from what I understand this depends on `CorrectnessMetric::name`, so does not belong here

File: giskard/rag/metrics/correctness.py [hunk 5]
```diff
@@ -55,46 +51,39 @@
...
-            except Exception as err:
-                raise LLMGenerationError(f"Error while evaluating the agent: {err}")
-
-        return {"correctness": pd.DataFrame(results).set_index("id")}
+                    ),
+                ],
+                temperature=0,
+            )
+            evaluation = json.loads(out.content, strict=False)
+            return evaluation
```
Comment (mattbit) [inline giskard/rag/metrics/correctness.py:84]: maybe worth checking that we return the correct thing

File: giskard/rag/metrics/ragas_metrics.py [hunk 6]
```diff
@@ -1,27 +1,28 @@
...
     from ragas.metrics import answer_relevancy, context_precision, context_recall, faithfulness
     from ragas.metrics.base import Metric as BaseRagasMetric
+    from ragas.run_config import RunConfig
+
 
 except ImportError as err:
-    logger.error(
+    raise ImportError(
         f"Package {err.name} is missing, it is required for the computation of RAGAS metrics. You can install it with `pip install {err.name}`."
     )
```
Comment (mattbit) [inline giskard/rag/metrics/ragas_metrics.py:27]: from err

**Review by mattbit — CHANGES_REQUESTED** [review]

File: giskard/datasets/metadata/text_metadata_provider.py (hunk 1 above)
Comment (pierlj) [reply giskard/datasets/metadata/text_metadata_provider.py:38]: I move it into `giskard/utils/language_detection.py`

**Review by mattbit — APPROVED** [review]

**Pull request merged**